### PR TITLE
Remove `:host` selectors in custom-styles if they are applying to the main document

### LINKED ISSUE #193: Rules like html, :host {...} don't have :host replaced correctly
I am stamping a `dom-module` that sort of looks like this:
`
<dom-module id="paper-item-styles">
  <template>
    <style>
      html. :host {
       ...
      }
    </style>
</template>
</dom-module>
`

And then importing it somewhere like:
`
<custom-style>
  <style is="custom-style" include="paper-item-styles"></style>
</custom-style>
`

And this works in Chrome and not FF (:host doesn't get transformed and the whole selector is invalid)

### PR DESCRIPTION
**Pull request by azakus**
Also backport :-webkit-any selector fixes from Polymer 1.x

Fixes #193

### CODE REVIEW

File: src/style-util.js [hunk 1]
```diff
@@ -291,4 +291,39 @@ export function gatherStyleText(element) {
...
+
+/**
+ * Split a selector separated by commas into an array in a smart way
+ * @param {string} selector
+ * @return {!Array<string>}
+ */
+export function splitSelectorList(selector) {
+  let parts = [];
+  let part = '';
+  for (let i = 0; i >= 0 && i < selector.length; i++) {
```
Comment (bicknellr) [inline src/style-util.js]: Would `i` ever be less than 0?

Comment (bicknellr) [reply src/style-util.js]: (from offline) `findMatchingParen` might return -1 and `i = end;` happens unconditionally after it.

File: tests/style-transformer.html [hunk 2]
```diff
@@ -99,6 +110,10 @@
     test(':host > *{}', function() {
       assert.equal(ast.rules[2].selector, 'x-foo > *.x-foo');
     });
+
+    test(':host() :-webkit-any()', function() {
+      assert.equal(ast.rules[3].selector, 'x-foo.bar .x-foo:-webkit-any(.foo, .bar).x-foo');
```
[the remark this answers is not in the document]
Comment (azakus) [reply tests/style-transformer.html]: Looks like `_transformCompoundSelector` is walking into the `:-webkit-any` selector. I'll have to work around this by removing spaces inside the `:-webkit-any()` block, and just not support complex selectors

File: src/style-transformer.js [hunk 3]
```diff
@@ -273,6 +273,40 @@ class StyleTransformer {
...
+  /**
+   * Replace MATCHES_REPLACMENT character with the given set of `:matches()`
+   * selectors.
+   *
+   * @param {string} selector
+   * @param {!Array<string>} matches
+   * @return {string}
+   */
+  _replaceMatchesPseudo(selector, matches) {
+    return selector.replace(MATCHES_REPLACEMENT, () => matches.pop());
```
Comment (bicknellr) [inline src/style-transformer.js]: This only replaces the first occurrence of `MATCHES_REPLACEMENT`.

File: src/style-transformer.js [hunk 4]
```diff
@@ -427,5 +473,7 @@ let CSS_CLASS_PREFIX = '.';
 let PSEUDO_PREFIX = ':';
 let CLASS = 'class';
 let SELECTOR_NO_MATCH = 'should_not_match';
+const MATCHES = /:(?:matches|any|-(?:webkit|moz)-any)/;
+const MATCHES_REPLACEMENT = '\u{1234}';
```
[the remark this answers is not in the document]
Comment (azakus) [reply src/style-transformer.js]: Fine by me

File: src/style-transformer.js (hunk 3 above)
Comment (azakus) [reply src/style-transformer.js]: Ah yeah, good catch!

File: src/style-transformer.js [hunk 5]
```diff
@@ -273,6 +273,40 @@ class StyleTransformer {
...
+   * @param {string} selector
+   * @return {{selector: string, matches: !Array<string>}}
+   */
+  _preserveMatchesPseudo(selector) {
+    /** @type {!Array<string>} */
+    const matches = [];
+    let match;
+    while ((match = selector.match(MATCHES))) {
+      const start = match.index;
+      const end = StyleUtil.findMatchingParen(selector, start);
```
[the remark this answers is not in the document]
Comment (azakus) [reply src/style-transformer.js]: I'm trying not to interpret user CSS, for maximum compatibility. However, unmatched parens is a bad thing no matter what. I suppose we could throw here.

[the remark this answers is not in the document]
Comment (azakus) [reply src/style-transformer.js]: Decided to throw the string of the selector, saying it has an unmatched paren.

File: tests/deferred-apply.html [hunk 6]
```diff
@@ -52,24 +52,25 @@
...
           const el = document.querySelector('x-foo');
           assert.equal(getComputedStyle(el).getPropertyValue('background-color').trim(), 'rgb(0, 0, 255)');
         });
-        if (document.readyState === 'complete' || document.readyState === 'interactive') {
+        if (document.readyState === 'complete') {
           resolveFn();
         } else {
+          window.addEventListener('load', resolveFn);
           window.addEventListener('DOMContentLoaded', () => {
+            window.removeEventListener('load,', resolveFn);
```
Comment (bicknellr) [inline tests/deferred-apply.html]: `load,` -> `load`

Comment (azakus) [reply tests/deferred-apply.html]: 🤦‍♂️

**Pull request merged**